clang/test/3C/multiple_tu.c

```diff
@@ -0,0 +1,22 @@
+// RUN: rm -rf %t
+// RUN: mkdir %t && cd %t
+// RUN: python -c 'import sys, json; json.dump([{"arguments": ["clang", "-c", "%s"], "directory": "%S", "file": "%s"}]*2, sys.stdout)' > compile_commands.json
+// RUN: 3c -dump-stats -p %t -base-dir=%S %s | FileCheck -match-full-lines %s
+// RUN: python -c 'import sys, json; exit(any(e["Name"].startswith("ImplicitCastExpr") for e in json.load(sys.stdin)["RootCauseStats"]))' < PerWildPtrStats.json
+
+// The compilation database used for this test includes two entries for this
+// file, causing 3C to process it twice. In issue #661, this caused type
+// argument instantiation to fail.
+
+// This issue also made an erroneous entry appear in the root cause output. The
+// root cause did not appear in the -warn-root-cause warnings. It was only
+// present in the root cause statistics json output. The json is used to
+// generate the output for 3c-wrap root_cause, so the error appeared there as well.
```

Comment:
That's interesting.  Ideally, our regression tests would ensure the correctness of the JSON output on a wider range of code constructs than just this one.  I hoped that testing the `-warn-all-root-cause` output via the diagnostic verifier would check the correctness of the same data that goes into the JSON, making it unnecessary to also check the JSON directly.  But if not, maybe we should change the entire `root_cause.c` test to verify the JSON.  Am I missing some easy way out, or shall I file a follow-up issue?

Reply:
They're two reasons that a root cause might show up in the JSON but not in the diagnostics.

1. The simple case (which is not what happened here) is that a diagnostic won't be emitted if they're isn't a source location for the root cause. 
2. What's happening here is that a new constraint was generated between the root cause diagnostics and JSON statistic output. This shouldn't ever happen, but there's isn't any mechanism in place to ensure that it doesn't. A recurring source of issue has been that the `CastPlacementVisitor` uses `getExprConstraintVars` to get constraint variable for parameter expressions. `getExprConstraintVars` will sometimes generate new constraints and variables. We tried to avoid this by introducing the `ExprConstraintVars` map to allocating new ConstraintVariables for the same expression every time `getExprConstraintVars` is called for it. It'd not entirely clear to me why this didn't work in this case, and what relation it has to the type variable insertion bug.

To try to avoid this type of bug in the future, we should think of a way to enforce that no new constraints or variables should be created after the constraint generation pass. A simple way to achieve this might be a boolean flag and some well placed asserts. A better solution that would require more engineering effort would be to use C++ `const` qualifiers to enforce this at compile time.

As for testing: since `3c-wrap root_cause` comes from JSON output, I think we should test it directly. Perhaps it would be enough to check for an exact match been the generated JSON and some expected JSON file (probably using python to do `json.loads("expected.json") == json.loads("actual.json")`). This would be overly delicate, but would detect changes to the root cause output that don't appear in the warnings.